core/features/abundance_task.py

```diff
@@ -0,0 +1,142 @@
+        2. Filter out incomplete-data site observations.
+        3. Generate dummy variables for land-use, combined land-use and
+        intensity, as well as some special case dummies from BII.
+        4. Transform continuous variables (road and population density)
+        5. Calculate mean values for population and road density.
+        6. For each groupby key (different levels of granularity), calculate
+        and scale abundance per site.
+        """
+        logger.info("Initiating feature pipeline for abundance.")
+        start = time.time()
+
+        # Read PREDICTS data and keep only columns that we need
+        df = pl.read_parquet(self.combined_data, columns=self.cols_to_keep)
+
+        # Filter out cases where land-use or intensity is missing
+        df = filter_out_insufficient_data_studies(df)
+
+        # Create dummy variables for land-use related columns
+        df = create_land_use_dummies(df, col_order=self.land_use_col_order)
+        df = combine_land_use_intensity_columns(df, col_order=self.lui_col_order)
+        df = group_land_use_types_and_intensities(
+            df, col_order=self.secondary_veg_col_order
+        )
+
+        # Generate various transformations for all continuous variables
+        df, new_cols = transform_continuous_covariates(
+            df, continuous_vars=self.continuous_vars
+        )
+
+        # Calculate mean values for population and road density, per resolution
+        df = calculate_study_mean_densities(
+            df, cols_to_incl=self.continuous_vars + new_cols
+        )
+
+        # List of columns to group by. For each deeper level in the taxonomic
+        # hierarchy, another column is added to the base list
+        groupby_cols = ["SS", "SSB", "SSBS"]
+
+        for i, path in enumerate(self.abundance_data):
+            df_abund = calculate_scaled_abundance(df, groupby_cols=groupby_cols)
+
+            # Get the first instance of each SSBS for the specified covariates
+            # Drop columns that relate to individual taxon measurements
+            # Drop columns with more granular species info than the grouping
+            df_first = df.group_by("SSBS").agg(pl.first(df.columns))
+            df_first = df_first.drop(
+                [
+                    "Taxon_name_entered",
+                    "Measurement",
+                    "Effort_corrected_measurement",
+                ]
+            )
+            df_first = df_first.drop(self.taxonomic_levels[i:])
+            df_res = df_abund.join(
+                df_first,
+                on=groupby_cols,
+                how="left",
+            )
+
+            # Save output file for this level of aggregation
+            df_res.write_parquet(path)
+
+            # Update the level used for grouping
+            if i < len(self.taxonomic_levels):
+                groupby_cols.append(self.taxonomic_levels[i])
+            else:
+                break
```

Comment:
There are 6 elements in the list of data paths, going from all species down to family. The list of taxonomic levels include 5 elements, since "all species" is not included. So this is basically a bit of a hacky way to manage the disrepancy between the number of elements in these lists. Should be better commented.

Reply:
Oh yea, of course. Sorry, my brain did a dyslexic spin on the conditional.